pkg/controller/sync/aggregator/aggregator_test.go

```diff
@@ -0,0 +1,237 @@
+						Group:   "group2",
+						Version: "v1",
+						Kind:    "Kind2",
+					}: {},
+				},
+			},
+			expectRev: map[schema.GroupVersionKind]map[Key]struct{}{
+				{
+					Group:   "group1",
+					Version: "v1",
+					Kind:    "Kind1",
+				}: {
+					{
+						Kind: syncset,
+						Name: "foo",
+					}: {},
+				},
+				{
+					Group:   "group2",
+					Version: "v1",
+					Kind:    "Kind2",
+				}: {
+					{
+						Kind: syncset,
+						Name: "foo",
+					}: {},
+				},
+			},
+		},
+		{
+			name: "add two keys and GVKs",
+			keyGVKs: []upsertKeyGVKs{
+				{
+					key: Key{
+						Kind: syncset,
+						Name: "foo",
+					},
+					gvks: []schema.GroupVersionKind{
+						{
+							Group:   "group1",
+							Version: "v1",
+							Kind:    "Kind1",
+						},
+						{
+							Group:   "group2",
+							Version: "v1",
+							Kind:    "Kind2",
+						},
+					},
+				},
+				{
+					key: Key{
+						Kind: configsync,
+						Name: "foo",
+					},
+					gvks: []schema.GroupVersionKind{
+						{
+							Group:   "group1",
+							Version: "v1",
+							Kind:    "Kind1",
+						},
+						{
+							Group:   "group2",
+							Version: "v1",
+							Kind:    "Kind2",
+						},
+					},
+				},
+			},
+			expectAdded: true,
+			expectData: map[Key]map[schema.GroupVersionKind]struct{}{
+				{
+					Kind: syncset,
+					Name: "foo",
+				}: {
+					{
+						Group:   "group1",
+						Version: "v1",
+						Kind:    "Kind1",
+					}: {},
+					{
+						Group:   "group2",
+						Version: "v1",
+						Kind:    "Kind2",
+					}: {},
+				},
+				{
+					Kind: configsync,
+					Name: "foo",
+				}: {
+					{
+						Group:   "group1",
+						Version: "v1",
+						Kind:    "Kind1",
+					}: {},
+					{
+						Group:   "group2",
+						Version: "v1",
+						Kind:    "Kind2",
+					}: {},
+				},
+			},
+			expectRev: map[schema.GroupVersionKind]map[Key]struct{}{
+				{
+					Group:   "group1",
+					Version: "v1",
+					Kind:    "Kind1",
+				}: {
+					{
+						Kind: syncset,
+						Name: "foo",
+					}: {},
+					{
+						Kind: configsync,
+						Name: "foo",
+					}: {},
+				},
+				{
+					Group:   "group2",
+					Version: "v1",
+					Kind:    "Kind2",
+				}: {
+					{
+						Kind: syncset,
+						Name: "foo",
+					}: {},
+					{
+						Kind: configsync,
+						Name: "foo",
+					}: {},
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			agg := NewGVKAggregator()
+
+			for _, keyGVKs := range tt.keyGVKs {
+				added, err := agg.UpsertWithValidation(keyGVKs.key, keyGVKs.gvks)
+
+				if tt.expectErr {
+					assert.Error(t, err, "expected an error but none occurred")
+					return
+				} else {
+					require.Equal(t, tt.expectAdded, added, "returned value did not match expected")
+				}
+			}
+
+			// these are open box tests for the underlying implementation of the GVKAggregator
+			require.Equal(t, tt.expectData, agg.(*bidiGVKAggregator).store, "data map did not match expected")            //nolint:forcetypeassert
```

Comment:
`require` is new for me.  Is this used elsewhere in Gatekeeper?

Reply:
yes we(I) use `stretchr/testify` in gatekeeper testing. I added the dep a while ago to help w test verbosity.